### DISCUSSION TITLE
Generic Validation Errors

### DISCUSSION THREAD

**Discussion by twiclo** [MichalLytek/type-graphql#1677] [Q&A]
Stripped down code:
```
import 'reflect-metadata';
import express, { ErrorRequestHandler, Request as ExpressRequest, Response } from 'express';
import { createHandler } from 'graphql-http/lib/use/express';
import { buildSchema } from 'type-graphql';
import { ContactResolver } from './app/resolvers/contact';
import { AddressResolver } from './app/resolvers/address';

const app = express();

app.use(express.json());

db.sequelize.sync({ force: true }).then(async () => {
	const schema = await buildSchema({
		resolvers: [
			ContactResolver,
			AddressResolver,
		],
		// I saw in an issue from years ago that adding this might help. Didn't seem to do anything
		validate: { forbidUnknownValues: false },
	});


	app.use("/graphql", (req, res, next) => {
		if (req.method === "POST" && req.body && req.body.operationName != 'IntrospectionQuery') {
    }
		next();
	}, createHandler({
		schema: schema
	}));

	app.listen(config.app.port, async () => {
		console.log(`\nRunning in ${config.app.dev ? "DEV" : "PROD"} mode on port ${config.app.port}\n`);
	});
})
```
Example input type:
```
@InputType()
class NewContact extends Serialize {
	@Field()
	firstName!: string;

	@Field()
	lastName!: string;

	@Field()
	@IsPhoneNumber("US")
	phone!: string;
}
```
Here's what I get when I try to submit a bad phone number on a field with 
```
{
  "errors": [
    {
      "message": "Argument Validation Error",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "createContact"
      ]
    }
  ],
  "data": null
}
```

I read in the docs that apollo server can automatically convert the validation error to the correct error and handles sending it back just fine but I'd like to avoid apollo server if possible. Do I need to write some sort of middleware for this? Is that all the apollo package is doing? I know that detailed validation errors are there. I've seen class-validator output them when working with it directly. Any guidance would be appreciated.

**Comment by MichalLytek** [comment]
Your goal is to hide validation error? How would you then know what was the reason of form submit error, etc.?

**Reply by twiclo** [reply]
My goal is the opposite. That validation error I pasted is too vague. I want to see which field failed and for what reason. Just like what is shown in the docs if you're using the apollo server.

**Reply by MichalLytek** [reply]
So what you're using to expose graphql schema via http, if not apollo server?

**Reply by twiclo** [reply]
Correct. It's just express.

**Reply by MichalLytek** [reply]
https://www.npmjs.com/package/express-graphql
```
customFormatErrorFn: An optional function which will be used to format any errors produced by fulfilling a GraphQL operation. If no function is provided, GraphQL's default spec-compliant [formatError](https://github.com/graphql/graphql-js/blob/master/src/error/formatError.js) function will be used.
```

**Reply by twiclo** [reply]
Well you can see I don't provide that option when setting it up so why doesn't it default to the default? I don't expect you to have the solution here but if you can point me in the right direction I'd appreciate it.

**Reply by MichalLytek** [reply]
The direction is to setup your graphql->http layer in a way that will expose additional errors information that are hidden by default for security reasons.

**Reply by twiclo** [reply]
Not sure what you're talking about. This is from the [type-graphql docs](https://typegraphql.com/docs/validation.html):

>By default, the apollo-server package from the [bootstrap guide](https://typegraphql.com/docs/bootstrap.html) will format the error to match the GraphQLFormattedError interface. So when the ArgumentValidationError occurs, the client will receive this JSON with a nice validationErrors property inside of extensions.exception:
```
{
  "errors": [
    {
      "message": "Argument Validation Error",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": ["addRecipe"],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "validationErrors": [
            {
              "target": {
                "title": "Lorem ipsum dolor sit amet, Lorem ipsum dolor sit amet"
              },
              "value": "Lorem ipsum dolor sit amet, Lorem ipsum dolor sit amet",
              "property": "title",
              "children": [],
              "constraints": {
                "maxLength": "title must be shorter than or equal to 30 characters"
              }
            }
          ],
          "stacktrace": [
            "Error: Argument Validation Error",
            "    at Object.<anonymous> (/type-graphql/src/resolvers/validate-arg.ts:29:11)",
            "    at Generator.throw (<anonymous>)",
            "    at rejected (/type-graphql/node_modules/tslib/tslib.js:105:69)",
            "    at processTicksAndRejections (internal/process/next_tick.js:81:5)"
          ]
        }
      }
    }
  ],
  "data": null
}
```
So by default I would be getting a useful validation error talking about how the length is too long. Apollo is set up to enable this `GraphQLFormattedError` but I'm not sure how to get that with Express.

**Reply by twiclo** [reply]
Could you please let me know how I can maybe just get the detailed errors printing to the console?
